src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl

```diff
@@ -201,244 +196,552 @@ function non_orographic_gravity_wave_tendency!(
         sqrt(abs(ᶜbuoyancy_frequency)),
     ) # to avoid small numbers
 
+    # prepare physical uv input variables for gravity_wave_forcing()
+    ᶜu = Geometry.UVVector.(Y.c.uₕ).components.data.:1
+    ᶜv = Geometry.UVVector.(Y.c.uₕ).components.data.:2
+
     if model_config isa SingleColumnModel
         # source level: the index of the level that is closest to the source height
 
-        Operators.column_mapreduce!(
-            min_distance_reduce,
-            source_level_z,
-            ᶜz,
-            ᶜlevel,
-        ) do z, level
-            (abs.(z .- gw_source_height), level)
+        input = Base.Broadcast.broadcasted(tuple, ᶜρ, ᶜz, ᶜu, ᶜv, ᶜlevel)
+        Operators.column_reduce!(
+            source_ρ_z_u_v_level,
+            input;
+        ) do (ρ_prev, z_prev, u_prev, v_prev, level_prev), (ρ, z, u, v, level)
+            if abs(z_prev - gw_source_height) >= abs(z - gw_source_height)
+                return (ρ, z, u, v, level, gw_source_height)
+            else
+                return (ρ_prev, z_prev, u_prev, v_prev, level_prev)
+            end
```

Comment:
Also, this doesn't seem right, the number of elements returns depends on the values of the data. I don't think that we need to return `gw_source_height` in the `if`-branch

Reply:
Can we, instead, do this as something like:

```julia
        input = Base.Broadcast.broadcasted(tuple, ᶜz, ᶜlevel)
        Operators.column_reduce!(
            source_z_level,
            input;
        ) do (z_prev, level_prev), (z, level)
            if abs(z_prev - gw_source_height) >= abs(z - gw_source_height)
                return (z, level)
            else
                return (z_prev, level_prev)
            end
        end
        ᶜz_source = source_z_level.:1
        ᶜlevel = source_z_level.:2
        ᶜρ_source = Spaces.level.(ᶜρ, ᶜlevel[1])
        ᶜu_source = Spaces.level.(ᶜu, ᶜlevel[1])
        ᶜv_source = Spaces.level.(ᶜv, ᶜlevel[1])
```
?

Reply:
And same for below?

Reply:
I think we cannot do that, since the source_level may be different in different columns.

Reply:
> Also, this doesn't seem right, the number of elements returns depends on the values of the data. I don't think that we need to return `gw_source_height` in the `if`-branch

Oh sorry, I forget to remove the  `gw_source_height`, we do not need it here. I will remove it!

Reply:
> I think we cannot do that, since the source_level may be different in different columns.

Why is that a problem? Won't `source_z_level` be a field of `Int`s that can then be used to compute the other variables?
```julia
        ᶜz_source = source_z_level.:1
        ᶜlevel = source_z_level.:2
        ᶜρ_source = Spaces.level.(ᶜρ, ᶜlevel[1])
        ᶜu_source = Spaces.level.(ᶜu, ᶜlevel[1])
        ᶜv_source = Spaces.level.(ᶜv, ᶜlevel[1])
```
Or, maybe I'm misinterpreting what `column_reduce!` is doing.

Reply:
Yes, here ᶜlevel is a surface field and ᶜρ is a space field, we want to get ᶜρ_source 、as a surface field. This code ᶜρ_source = Spaces.level.(ᶜρ, ᶜlevel[1]) will throw an error since we cannot directly getindex of a field. I remove '[1]' but it still cannot work.

Reply:
Maybe we can connect about this, I think that there should be a way to make this work, and I think it's a lot simpler

Reply:
I think what we have is fine for now, we can revisit this later if we need.